### ISSUE TITLE
Ability to configure TTS engines in the UI

### ISSUE THREAD

**Issue by bertfrees** [daisy/pipeline-ui#151]
- [x] Some engines need to be configured before they can work. Examples are the cloud based engines from Google and Microsoft. There should be a config panel for entering the [available properties](http://daisy.github.io/pipeline/Get-Help/User-Guide/Text-To-Speech/) for each engine. A nice and simple way to present the settings could be to have one panel per available engine. Each engine would have a status to indicate whether it is working or not. (The GUI can check this using the API to retrieve voices.) The settings would be nicely hidden for unconfigured or disabled engines. Upon enabling an engine, the user is asked to fill in the required properties. If the properties are not filled in correctly, the engine is not enabled. The user can also disable engines that are configured correctly or that don't need configuration (e.g. the engines native to Windows and macOS).

- [x] Scripts with TTS support have a "Text-to-speech configuration file". For usability it would be nicer if this configuration would be integrated in the UI somehow. [#154](https://github.com/daisy/pipeline-ui/pull/154) adds a config panel for selecting preferred voices.

- [x] A nice addition to the voices config panel could be a way to have a live preview of the voices (blocked on https://github.com/daisy/pipeline-modules/issues/89).

**Comment by ways2read** [comment]
I guess "the cloud based engines always work" isn't true if there is no internet connection.

**Comment by bertfrees** [comment]
Some idea's from yesterday's team meeting:

@rdeltour:

> One thing that c/b nice is to have a way to have a live preview of the TTS from a config page in the UI.

@bertfrees:

> I want to eliminate the "Text-to-speech configuration file" options, and replace it with the following:
> - a dedicated option to specify CSS style sheets (in addition to the possibility to attach style sheets to the input)
> - a dedicated option to specify lexicons (in addition to the possibility to attach lexicons to the input)
> - dedicated options for certain TTS properties
>   - `org.daisy.pipeline.tts.log`: done
>   - `org.daisy.pipeline.tts.mp3.bitrate`: to do
>   - `org.daisy.pipeline.tts.lame.cli.options`: has been [deprecated](https://github.com/daisy/pipeline-modules/commit/8b28a8800dd7e23a1428c5233f5fae2a4e92441f)
> - it should not be possible anymore to set other TTS properties dynamically (note that [`org.daisy.pipeline.tts.host.protection` has already been deprecated](https://github.com/daisy/pipeline/issues/507))
> - if needed, the per-job voice configuration could be replaced by a system wide configuration

**Comment by bertfrees** [comment]
This issue is not quite fixed yet I think. https://github.com/daisy/pipeline-ui/pull/154 fixes only a part of it, namely selecting preferred voices.

**Comment by marisademeglio** [comment]
> This issue is not quite fixed yet I think. #154 fixes only a part of it, namely selecting preferred voices.

Can you create more issues for what still needs to be done? Or is it more of keeping everyone's ideas around?

**Comment by bertfrees** [comment]
It is more a collection of ideas. We can create some new issue with concrete things to do.

**Comment by bertfrees** [comment]
On second thought, the first comment in this issue sums it up pretty well I think. Instead of creating a new issue with more or less the same in it, I'm gonna reword this one, and convert it into a list of tasks.

**Comment by marisademeglio** [comment]
Will there be an API to describe engines' properties? Or should I hardcode it based on the [engine configuration](http://daisy.github.io/pipeline/Get-Help/User-Guide/Text-To-Speech/ ) docs? 
 
Voice preview will come from the API too, right? Is that ready?

**Comment by bertfrees** [comment]
I think hardcoding the properties makes the most sense for now. But an API for the engines can definitely be useful too. Let's keep the idea.

Voice preview will come from the API, yes. I'm not sure yet what the API should look like though. I guess it could also be a general purpose "speak" command, that could even accept SSML. That wouldn't be so hard to do. (A while back I already wrote a [mock of the Google TTS API](https://github.com/daisy/pipeline-modules/blob/b5d5013/tts/tts-mocks/src/main/java/org/daisy/pipeline/tts/mock/impl/MockGoogle.java) that dispatches to the available TTS engines.)

**Comment by marisademeglio** [comment]
Added credential fields for Azure and Google voices in 4d97c87be89cf94428d999aeef0016ac28f0c63a

Verifying the credentials is a new issue: https://github.com/daisy/pipeline-ui/issues/164

From [this convo](https://daisy-dev.slack.com/archives/C064GB8U9/p1696263746999069?thread_ts=1696262507.250939&cid=C064GB8U9) we have now implemented all the engine settings for our current goal

**Comment by marisademeglio** [comment]
Noting here that we also got a feature request from a tester:
"When selecting between voices, offer a preview."

**Comment by marisademeglio** [comment]
Is there anything left in the first task above that is still relevant? We have designed the settings dialog in a different way based on other convos about engine properties that we wanted to support.

And the "voice preview" task is still pending engine implementation.

**Comment by bertfrees** [comment]
> Is there anything left in the first task above that is still relevant?

The way the settings dialog looks now is great!

It's very minor, but one thing that would be nice is if the status (connected or disconnected) would somehow be made more clear. I don't know how though.

By the way, this note at the top:

> After configuring these engines with the required credentials, they will be available under 'Voices'. Save and reopen the settings dialog to see changes.

Is it really needed? It seems the voices are updated without closing and reopening the settings.

**Comment by marisademeglio** [comment]
True, that wording can be simplified as the changes now are effective immediately.

How is this for a slightly clearer connected/disconnected status?

<img width="912" alt="Screenshot 2024-04-17 at 09 35 56" src="https://github.com/daisy/pipeline-ui/assets/734730/459421e5-4041-48a2-996c-5b3885103e36">

**Comment by bertfrees** [comment]
Yes, I also thought of emphasizing it visually like that. That is indeed slightly better, however I don't know whether that fundamentally changes anything? 'Cause it will just be decoration, right?

Perhaps the engines could be grouped by connection status? There would be two main headings with the connection status, under which the subheadings "Azure" and "Google" would go. The main headings wouldn't need to be visible for sighted users, the sections could be indicated some other way.

Just thinking out loud. As I said, it is already good the way it is now.

**Comment by marisademeglio** [comment]
Ok I will commit this since it seems better.

We can keep this thread going for ideas. I don't like grouping it by connection status because then doesn't it get reordered when the status changes? That's visually disruptive and probably bad accessibility.

**Comment by bertfrees** [comment]
Yes, that's probably true.

**Comment by marisademeglio** [comment]
Voice preview team discussion:

https://daisy-dev.slack.com/archives/C064GB8U9/p1719243930802729
https://daisy-dev.slack.com/archives/C064GB8U9/p1719244054736299

**Comment by bertfrees** [comment]
Gonna close this one as the voice preview was the latest thing on the list.
